### ISSUE TITLE
rTorrent SCGI doesnt seem to work as a downloader

### ISSUE THREAD

**Issue by Arathen** [nosmokingbandit/Watcher3#20]
I'm using rTorrent SCGI as a downloader but sending torrents and magnets to it from Watcher doesn't seem to actually work. I dont think its a problem with my rTorrent config because im using it successfully with other searchers (Medusa and Mylar).

Logs as follows:

```
INFO 2017-04-17 14:53:04,101 core.snatcher.snatch_torrent: Sending magnet to rTorrent.
ERROR 2017-04-17 14:53:04,102 cherrypy.error.1973115664.error: [17/Apr/2017:14:53:04] HTTP
Traceback (most recent call last):
  File "/home/watcher/lib/cherrypy/_cprequest.py", line 670, in respond
    response.body = self.handler()
  File "/home/watcher/lib/cherrypy/lib/encoding.py", line 220, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/home/watcher/lib/cherrypy/_cpdispatch.py", line 60, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/watcher/core/ajax.py", line 307, in manual_download
    return json.dumps(self.snatcher.snatch(data))
  File "/home/watcher/core/snatcher.py", line 129, in snatch
    response = self.snatch_torrent(data)
  File "/home/watcher/core/snatcher.py", line 277, in snatch_torrent
    response = rtorrent.rTorrentSCGI.add_torrent(data)
  File "/home/watcher/core/downloaders/rtorrent.py", line 55, in add_torrent
    connected = rTorrentSCGI.test_connection(conf)
  File "/home/watcher/core/downloaders/rtorrent.py", line 28, in test_connection
    client = rtorrent.SCGIServerProxy(url)
  File "/home/watcher/lib/rtorrent.py", line 142, in __init__
    type, uri = urllib.splittype(uri)
AttributeError: module 'urllib' has no attribute 'splittype'
INFO 2017-04-17 14:53:04,184 cherrypy.error.1973115664.error: [17/Apr/2017:14:53:04] HTTP
Request Headers:
  ACCEPT: */*
  REFERER: http://obfuscated:9090/watcher/
  Remote-Addr: 192.168.0.1
  ORIGIN: http://obfuscated:9090
  X-REQUESTED-WITH: XMLHttpRequest
  ACCEPT-ENCODING: gzip, deflate
  COOKIE: much obfuscated data
  Content-Length: 107
  USER-AGENT: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36
  Content-Type: application/x-www-form-urlencoded; charset=UTF-8
  ACCEPT-LANGUAGE: en-US,en;q=0.8
  HOST: obfuscated:9090
  CONNECTION: keep-alive
```

Clicking the "Test Connection" button times out with the same error in the logs.

rTorrent/libtorrent versions are 0.9.6/0.13.6 which is running on the same host as Watcher, which is master commit 31e50c97912be0ba00e6421b9d10526a4ee7edb8 all on Ubuntu 16.04.2 armhf on a Pi2.

EDIT: Package versions are:

ii  python3                      3.5.1-3                                        armhf
ii  python3-urllib3          1.13.1-2ubuntu0.16.04.1             all   

Let me know if theres any other info you need.

Cheers :)

**Comment by nosmokingbandit** [comment]
I don't have the time or patience to install rTorrent again, but 72d8941faa2878b97a8175080881c50db93572f2 should take care of it.

**Comment by Arathen** [comment]
This seems to work as a fix for the immediate problem (thank you!), but then exposes another problem straight afterwards. It looks like just another hangover from the Python3 migration.

```
ERROR 2017-04-18 10:08:42,787 core.downloaders.rtorrent.test_connection: rTorrent connection test failed.
Traceback (most recent call last):
  File "/home/watcher/core/downloaders/rtorrent.py", line 31, in test_connection
    client.system.time()
  File "/usr/lib/python3.5/xmlrpc/client.py", line 1092, in __call__
    return self.__send(self.__name, args)
  File "/home/watcher/lib/rtorrent.py", line 174, in __request
    verbose=self.__verbose
  File "/usr/lib/python3.5/xmlrpc/client.py", line 1134, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/home/watcher/lib/rtorrent.py", line 91, in single_request
    header = '\x00'.join(('%s\x00%s' % item for item in headers.iteritems())) + '\x00'
AttributeError: 'dict' object has no attribute 'iteritems'
```

I fixed this one locally just by changing "iteritems" to "items" as per https://wiki.python.org/moin/Python3.0 but that then exposed a third problem as follows:

```
ERROR 2017-04-18 10:23:33,075 core.downloaders.rtorrent.test_connection: rTorrent connection test failed.
Traceback (most recent call last):
  File "/home/watcher/core/downloaders/rtorrent.py", line 31, in test_connection
    client.system.time()
  File "/usr/lib/python3.5/xmlrpc/client.py", line 1092, in __call__
    return self.__send(self.__name, args)
  File "/home/watcher/lib/rtorrent.py", line 174, in __request
    verbose=self.__verbose
  File "/usr/lib/python3.5/xmlrpc/client.py", line 1134, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/home/watcher/lib/rtorrent.py", line 103, in single_request
    socket.SOCK_STREAM)
  File "/usr/lib/python3.5/socket.py", line 732, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known
```

I've not had a chance to pick this one apart yet, sorry.

I hear you about rtorrent. I hate it. Its the worst piece of crap i've seen for a long time as far as configuration and usability goes. BUT, i'm seeding over 1000 torrents on a Raspberry Pi 2, and no other client comes close to the small footprint i need for that, so i'm kinda stuck with it :(

Would you prefer separate issues for these ones?

Thanks again for the help and continued support. Its much appreciated.

**Comment by nosmokingbandit** [comment]
I didn't write the rTorrent module, so bear with me while we get it hammered out.

34fe8e155acad35f63066599c8045f360168b28f should take care of the iter call. I didn't see anything else that looked invalid, but its hard to know without testing.

We'll keep everything in this issue until rTorrent works again for you.

**Comment by Arathen** [comment]
Sorry - no luck. I've updated and tested, but its still throwing this:

```
ERROR 2017-04-19 23:53:28,135 core.downloaders.rtorrent.test_connection: rTorrent connection test failed.
Traceback (most recent call last):
  File "/home/watcher/core/downloaders/rtorrent.py", line 35, in test_connection
    client.system.time()
  File "/usr/lib/python3.5/xmlrpc/client.py", line 1092, in __call__
    return self.__send(self.__name, args)
  File "/home/watcher/lib/rtorrent.py", line 166, in __request
    response = self.__transport.request(self.__host, self.__handler, request, verbose=self.__verbose)
  File "/usr/lib/python3.5/xmlrpc/client.py", line 1134, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/home/watcher/lib/rtorrent.py", line 102, in single_request
    addrinfo = socket.getaddrinfo(host, port, socket.AF_INET, socket.SOCK_STREAM)
  File "/usr/lib/python3.5/socket.py", line 732, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known
```
Let me know if theres any more info or testing you need from me.

Cheers.

**Comment by nosmokingbandit** [comment]
I'll just have to install rTorrent again. Maybe this time it won't take all afternoon to get it running....

**Comment by nosmokingbandit** [comment]
I just pushed a bunch of commits that should fix this, specifically c22f98ad3d14bb1c0c83eb1ba1d2e2f9a5aa95f5.

Using scgi connection I was able to send torrents and set labels, but I won't close this until I hear from you.

The problem is the way urllib.parse.urlparse works. It can't separate the port and hostname from 'localhost:5000', it requires 'scgi://localhost:5000' in order to split it correctly.

**Comment by Arathen** [comment]
We aren't there yet, but we're definitely getting closer :) I'm not sure how mine is different from yours, but this is the error I get...
```
ERROR 2017-04-20 12:14:46,828 core.downloaders.rtorrent.test_connection: rTorrent connection test failed.
Traceback (most recent call last):
  File "/home/watcher/core/downloaders/rtorrent.py", line 31, in test_connection
    client.system.time()
  File "/usr/lib/python3.5/xmlrpc/client.py", line 1092, in __call__
    return self.__send(self.__name, args)
  File "/home/watcher/lib/rtorrent.py", line 165, in __request
    response = self.__transport.request(self.__host, self.__handler, request, verbose=self.__verbose)
  File "/usr/lib/python3.5/xmlrpc/client.py", line 1134, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/home/watcher/lib/rtorrent.py", line 111, in single_request
    return self.parse_response(sock.makefile())
  File "/home/watcher/lib/rtorrent.py", line 127, in parse_response
    response_header, response_body = re.split(r'\n\s*?\n', response_body, maxsplit=1)
ValueError: not enough values to unpack (expected 2, got 1)
```
My downloader config is:

rTorrent SCGI
Host & Port: scgi://localhost:5995
Label: Watcher
Add paused: unchecked

Thanks again :)

**Comment by nosmokingbandit** [comment]
That error is kind of odd. Without being able to replicate it, I'm not sure *exactly* what is going on, but it looks like rTorrent isn't sending headers (or is sending invalid headers), which is why it can't be split.

I've made a little change in b24f909427572eb54d8a6c9d383d088ebdf38c89 that is a bit more flexible, but that doesn't mean its going to get the correct info.

**Comment by Arathen** [comment]
Ok... i've just had a chance to update and test. SCGI is getting the magnet into rtorrent now, which is progress. For some reason though, the magnet adds in a "Stopped" state, with a Watcher label. When I manually started the download, rtorrent removed the magnet and readded the torrent, without the label. It also re-added it in stopped state again, which required a second manual start (along with a manual label assignment so post processing would work).

Logs are harder to provide now though, as everything looks successful from Watchers perspective. I just see :
```
INFO 2017-04-22 14:56:40,834 core.snatcher.snatch_torrent: Sending magnet to rTorrent.
INFO 2017-04-22 14:56:42,326 core.sqldb.update: Updating downloadid to 51601f3592749aeaa1def58fc9fe8e91158f6043 for col guid:51601f3592749aeaa1def58fc9fe8e91158f6043 in SEARCHRESULTS.
INFO 2017-04-22 14:56:42,891 core.library.searchresults: Marking guid 51601f3592749aeaa1def58fc9fe8e91158f6043 as Snatched.
INFO 2017-04-22 14:56:42,923 core.library.searchresults: Marking 51601f3592749aeaa1def58fc9fe8e91158f6043 as Snatched in SEARCHRESULTS.
INFO 2017-04-22 14:56:42,924 core.sqldb.update: Updating status to Snatched for col guid:51601f3592749aeaa1def58fc9fe8e91158f6043 in SEARCHRESULTS.
INFO 2017-04-22 14:56:42,959 core.library.searchresults: Successfully marked 51601f3592749aeaa1def58fc9fe8e91158f6043 as Snatched in SEARCHRESULTS.
INFO 2017-04-22 14:56:42,985 core.library.markedresults: Marking 51601f3592749aeaa1def58fc9fe8e91158f6043 as Snatched in MARKEDRESULTS.
INFO 2017-04-22 14:56:42,986 core.sqldb.update: Updating status to Snatched for col guid:51601f3592749aeaa1def58fc9fe8e91158f6043 in MARKEDRESULTS.
INFO 2017-04-22 14:56:43,023 core.library.markedresults: Successfully marked 51601f3592749aeaa1def58fc9fe8e91158f6043 as Snatched in MARKEDRESULTS.
INFO 2017-04-22 14:56:43,024 core.sqldb.get_movie_details: Retrieving details for tt2072233.
INFO 2017-04-22 14:56:43,054 core.sqldb.get_distinct: Getting distinct values for tt2072233 in SEARCHRESULTS
INFO 2017-04-22 14:56:43,081 core.library.movie_status: Setting MOVIES tt2072233 status to Snatched.
INFO 2017-04-22 14:56:43,082 core.sqldb.update: Updating status to Snatched for col imdbid:tt2072233 in MOVIES.
INFO 2017-04-22 14:56:43,120 core.snatcher.snatch_torrent: Successfully sent Sleepless.2017.1080p.BluRay.x264-DON[EtHD] to rTorrent.
```
Manually selecting a torrent instead of a magnet avoids the re-adding situation, but it still adds in a "Stopped" state with no label on it.

Basically, SCGI seems to be working the same as HTTPRPC now.

Cheers once again for the help.

**Comment by nosmokingbandit** [comment]
I'm hoping 00af58e119a481d30a1d32aab628e76878f87612 fixes this for good. rTorrent can be slow when adding torrents and requires a second request to set the label, so if Watcher tries to load a torrent and set the label too quickly rTorrent won't have loaded the torrent yet.

I'm still having trouble with certain torrents not loading at all though. The rTorrent log shows that it is getting the torrent, but it completely ignores it. I'm ok with blaming rTorrent for that.

**Comment by Arathen** [comment]
Ok... everything is working correctly now, with the exception of one thing.

Magnets add with a label, but then lose their label when rtorrent re-adds the torrent from the magnet.

Both magnets and torrents now add in a downloading state, which is awesome. And torrents add directly with their label correct and work perfectly from what i can see in testing. Its just magnet labels which aren't sticking.

Great work on this though. I cant imagine the number of hours you've spent picking code apart, and im immensely grateful.

**Comment by nosmokingbandit** [comment]
> Magnets add with a label, but then lose their label when rtorrent re-adds the torrent from the magnet.

This is a known issue with rTorrent. When rTorrent gets a magnet it downloads the actual .torrent data and reloads the download. But when it reloads the download it doesn't preserve custom attributes like labels or download paths.

There are a few issues on rTorrent's github page but the dev keeps blaming the snatching application rather than actually fixing the problem.


That error you posted shows that rTorrent isn't sending xml data back. Everything is working fine on my testing server. Try restarting rTorrent.
